### PR TITLE
Fix #12565 and Tpetra::MatrixMatrix::Add in general

### LINKED ISSUE #12565: Tpetra:  Using Add to sum into a fillComplete matrix results in a matrix full of zeros
## Bug Report
@trilinos/tpetra 

### Description
I want to do `C = A+B` to update the values in a preexisting (fillComplete) crs matrix `C`.
According to the documentation `Tpetra::MatrixMatrix::Add` supports this, and the following snippet from `Add` has specific lines for fillComplete `C`:

https://github.com/trilinos/Trilinos/blob/6779ba1c05c50ccb12d78d6b04db737d8e3e4a2f/packages/tpetra/core/ext/TpetraExt_MatrixMatrix_def.hpp#L1128-L1134

However, `sumIntoGlobalValues` which is called in the fillComplete case above does not seem to be compatible with fill complete matrices, as indicated below:
https://github.com/trilinos/Trilinos/blob/6779ba1c05c50ccb12d78d6b04db737d8e3e4a2f/packages/tpetra/core/src/Tpetra_CrsMatrix_def.hpp#L2656-L2659

So no addition is done and you get back a matrix `C` that was set to zero at the beginning of `Add`, which isn't all that helpful.

### PR DESCRIPTION
**Pull request by brian-kelley**
@trilinos/tpetra 
@trilinos/muelu 

## Motivation

**In Tpetra**
- Fix https://github.com/trilinos/Trilinos/issues/12565: in Add that takes C as output argument, resume fill on
  C before calling sumIntoGlobalValues. Then fillComplete it again
  before returning.
- Correct documentation for Add version that adds A into B.
  - say that B must not be fill complete (this was already asserted in
    the function, so won't break user code)
  - don't say 'currently not functional' because it works fine
    and is tested
- Split Add version that takes A,B,C into 2 overloads
  - Backwards compatible
  - Before, took RCP by value.
  - Now, one takes const RCP& (adding into existing C) and the other takes RCP& (adding
    into existing C, or creating C if it's null on input). The
    documentation is updated and needs no warning saying "this doesn't
    work if C is null"
- Add a test for A,B,C Add where C is null on input.

**In MueLu**
ParameterListInterpreter_PointCrs_vs_BlockCrs is failing after
fixing Tpetra::MatrixMatrix::Add. The test uses Add to compute
C = A-B to check if A and B are close to equal, but when Add was
broken it was just zeroing out C, so the test passed before. The lines
that check C's Frobenius norm against a tolerance have been commented out here,
otherwise it wouldn't make it through PR testing. The underlying thing that this tests needs
to be fixed by MueLu, and then those lines can be uncommented again. 

## Testing

Checked with printf debugging that Tpetra's MatrixMatrix_UnitTests exercise the three versions of Add.

### CODE REVIEW

File: packages/tpetra/core/ext/TpetraExt_MatrixMatrix_def.hpp [hunk 1]
```diff
@@ -1066,14 +1070,35 @@ void Add(
...
+    if(Aprime->getRowMap()->isSameAs(*Bprime->getRowMap())) {
+      LocalOrdinal numLocalRows = Aprime->getLocalNumRows();
+      Array<size_t> CmaxEntriesPerRow(numLocalRows);
+      for(LocalOrdinal i = 0; i < numLocalRows; i++) {
+        CmaxEntriesPerRow[i] = Aprime->getNumEntriesInLocalRow(i) + Bprime->getNumEntriesInLocalRow(i);
+      }
+      C = rcp (new crs_matrix_type (Aprime->getRowMap (), CmaxEntriesPerRow()));
+    }
+    else {
+      // Make sure Aprime and Bprime have global constants, before we ask for max entries per row.
```
Comment (jhux2) [inline packages/tpetra/core/ext/TpetraExt_MatrixMatrix_def.hpp:1099]: How is this enforced?

File: packages/tpetra/core/ext/TpetraExt_MatrixMatrix_def.hpp [hunk 2]
```diff
@@ -1125,16 +1152,48 @@ void Add(
...
   }
+  if(CwasFillComplete) {
+    // This version of fillComplete will reuse the domain
+    // and range maps from the previous fillComplete.
+    C->fillComplete();
+  }
+}
+
+// This version of Add takes C as const RCP&, so C must not be null on input (in this case,
+// it is allocated and constructed in this function). Otherwise, its behavior is identical
```
Comment (jhux2) [inline packages/tpetra/core/ext/TpetraExt_MatrixMatrix_def.hpp:1175]: This phrasing is a little weird to me.  Because C is a `const RCP&`, it can't be allocated in this version.

Comment (brian-kelley) [reply packages/tpetra/core/ext/TpetraExt_MatrixMatrix_def.hpp:1175]: Good catch, that was copied from the description of the ``RCP&`` version.

File: packages/tpetra/core/ext/TpetraExt_MatrixMatrix_def.hpp (hunk 1 above)
Comment (brian-kelley) [reply packages/tpetra/core/ext/TpetraExt_MatrixMatrix_def.hpp:1099]: It's asserted up on lines 1013 and 1016.

Comment (brian-kelley) [inline packages/tpetra/core/ext/TpetraExt_MatrixMatrix_def.hpp:1099]:
```suggestion
      // Note: above we checked that Aprime and Bprime have global constants, so it's safe to ask for max entries per row.
```

File: packages/tpetra/core/ext/TpetraExt_MatrixMatrix_def.hpp (hunk 2 above)
Comment (brian-kelley) [inline packages/tpetra/core/ext/TpetraExt_MatrixMatrix_def.hpp:1175]:
```suggestion
// This version of Add takes C as const RCP&, so C must not be null on input. Otherwise, its behavior is identical
```

**Comment by brian-kelley** [comment]
@jhux2 Thanks for the review, I updated those comments.

**Review by jhux2 — APPROVED** [review]
Lgtm, thanks.  Would you mind opening a MueLu issue about fixing the test, if you haven't already?

**Comment by brian-kelley** [comment]
@jhux2 Good idea, here you go: #12573
